### PR TITLE
VIA support for Planck Rev6

### PR DESCRIPTION
**Pull request by mixedfeelings**
## Description
Adds simple VIA support for Planck Rev6 with a default QWERTY layout 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by mixedfeelings** [comment]
@jackhumbert wanted me to tag him for approval of the vendor / product IDs before merging.

File: keyboards/planck/keymaps/via/keymap.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
```
Comment (zvecr) [inline keyboards/planck/keymaps/via/keymap.c:19]:
```suggestion

#define LOWER  FN_MO13
#define RAISE  FN_MO23

```

File: keyboards/planck/keymaps/via/keymap.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    * | Shift|   Z  |   X  |   C  |   V  |   B  |   N  |   M  |   ,  |   .  |   /  |Enter |
+    * |------+------+------+------+------+------+------+------+------+------+------+------|
+    * |      | Ctrl | Alt  | GUI  |Lower |    Space    |Raise | Left | Down |  Up  |Right |
+    * `-----------------------------------------------------------------------------------'
+    */
+    [0] =  LAYOUT_ortho_4x12 (
+        KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,  KC_T,   KC_Y,   KC_U,  KC_I,    KC_O,    KC_P,    KC_BSPC,
+        KC_ESC,  KC_A,    KC_S,    KC_D,    KC_F,  KC_G,   KC_H,   KC_J,  KC_K,    KC_L,    KC_SCLN, KC_QUOT,
+        KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,  KC_B,   KC_N,   KC_M,  KC_COMM, KC_DOT,  KC_SLSH, KC_SFTENT,
+        KC_NO,   KC_LCTL, KC_LALT, KC_LGUI, MO(1), KC_SPC, KC_SPC, MO(2), KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT
```
Comment (zvecr) [inline keyboards/planck/keymaps/via/keymap.c:37]:
```suggestion
        KC_NO,   KC_LCTL, KC_LALT, KC_LGUI, LOWER, KC_SPC, KC_SPC, RAISE, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT
```

File: keyboards/planck/keymaps/via/keymap.c [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    * |      |  F7  |  F8  |  F9  |  F10 |  F11 |  F12 |ISO ~ |ISO | | Home | End  |      |
+    * |------+------+------+------+------+------+------+------+------+------+------+------|
+    * |      |      |      |      |      |             |      | Next | Vol- | Vol+ | Play |
+    * `-----------------------------------------------------------------------------------'
+    */
+    [1] = LAYOUT_ortho_4x12 (
+        KC_TILD, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC, KC_CIRC, KC_AMPR,    KC_ASTR,    KC_LPRN, KC_RPRN, KC_BSPC,
+        KC_DEL,  KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_UNDS,    KC_PLUS,    KC_LCBR, KC_RCBR, KC_PIPE,
+        KC_TRNS, KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  S(KC_NUHS), S(KC_NUBS), KC_HOME, KC_END,  KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, MO(3),      KC_MNXT,    KC_VOLD, KC_VOLU, KC_MPLY
```
Comment (zvecr) [inline keyboards/planck/keymaps/via/keymap.c:55]:
```suggestion
        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,    KC_MNXT,    KC_VOLD, KC_VOLU, KC_MPLY
```

File: keyboards/planck/keymaps/via/keymap.c [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    * |      |  F7  |  F8  |  F9  |  F10 |  F11 |  F12 |ISO # |ISO / |Pg Up |Pg Dn |      |
+    * |------+------+------+------+------+------+------+------+------+------+------+------|
+    * |      |      |      |      |      |             |      | Next | Vol- | Vol+ | Play |
+    * `-----------------------------------------------------------------------------------'
+    */
+    [2] = LAYOUT_ortho_4x12 (
+        KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,   KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_BSPC,
+        KC_DEL,  KC_F1,   KC_F2,   KC_F3,   KC_F4,  KC_F5,   KC_F6,   KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC, KC_BSLS,
+        KC_TRNS, KC_F7,   KC_F8,   KC_F9,   KC_F10, KC_F11,  KC_F12,  KC_NUHS, KC_NUBS, KC_PGUP, KC_PGDN, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, MO(3),  KC_TRNS, KC_TRNS, KC_TRNS, KC_MNXT, KC_VOLD, KC_VOLU, KC_MPLY
```
Comment (zvecr) [inline keyboards/planck/keymaps/via/keymap.c:73]:
```suggestion
        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,KC_TRNS, KC_TRNS, KC_TRNS, KC_MNXT, KC_VOLD, KC_VOLU, KC_MPLY
```

File: keyboards/planck/config.h [hunk 5]
```diff
@@ -21,8 +21,8 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "config_common.h"
 
 /* USB Device descriptor parameter */
-#define VENDOR_ID       0xFEED
-#define PRODUCT_ID      0x6060
+#define VENDOR_ID       0x6F6C //ol
```
Comment (zvecr) [inline keyboards/planck/config.h:24]: @jackhumbert thoughts on what you want the VID/PID set to?

File: keyboards/planck/keymaps/via/keymap.c (hunk 2 above)
Comment (mixedfeelings) [reply keyboards/planck/keymaps/via/keymap.c:40]: Great suggestion, much more efficient!

**Comment by mixedfeelings** [comment]
@Erovia it seems like the latter option would be the easiest to implement as you would only need two JSONs (for now). Perhaps the PID could refer to matrix revision rather than actual product.

 @jackhumbert did you have any further thoughts on VID / PID?

Thanks for looking at this, everyone!

**Comment by Erovia** [comment]
In any case, 2 json should be enough.  
In the first case, the rev1-rev5,ez boards could share the same PID.

I kinda hate the idea of repeating the same thing 6 different places, but also hate the idea of having 1 exception. ¯\_(ツ)_/¯

**Comment by drashna** [comment]
Does VIA not read the `DEVICE_VER`?

**Comment by noroadsleft** [comment]
> Does VIA not read the `DEVICE_VER`?

It does not.

**Comment by mixedfeelings** [comment]
Thanks everyone, I guess that leaves me in a holding pattern until it is decided how to handle PIDs for the multiple Planck matrices.

**Comment by fauxpark** [comment]
Personally I'd like to see VIA take into account the `DEVICE_VER` value. This seems like the perfect use case for it.

**Comment by Erovia** [comment]
@Wilba6582 @olivia @yiancar
Could you please weigh in on this topic?

**Comment by jackhumbert** [comment]
The split between the versions happens at the microcontroller level (rev1-5 vs rev6 & ez), so I think we can use these values (all have the Vendor ID of `0xFEED`):

* rev1-5: `0xAE01`
* rev6 & ez `0xA4F9`

These were generated with the strings "Planck AVR" and "Planck ARM" using the (somewhat) standard command:

     echo -n "<PRODUCT>" | shasum | cut -c1-4 |tr '[:lower:]' '[:upper:]'

**Comment by mixedfeelings** [comment]
Good info, @jackhumbert, thank you. Do you have any thoughts on VendorID? I don't think we can use 0xFEED for VIA.

For @noroadsleft, @Erovia, @fauxpark: Does the micro controller level distinction get us around this particular block for the purpose of officially supporting VIA for planck?

**Comment by fauxpark** [comment]
I think we should probably also make a distinction between the rev6 (ARM) and EZ, which would be `0xC6CF`.

**Comment by drashna** [comment]
> For @noroadsleft, @Erovia, @fauxpark: Does the microcontroller level distinction get us around this particular block for the purpose of officially supporting VIA for planck?

It should, I think.  The AVR versions have no real differences, other than backlight/rgb support, and bootloader, IIRC.  Everything else is the same. 



> I think we should probably also make a distinction between the rev6 (ARM) and EZ, which would be `0xC6CF`.

For the Planck EZ, would prefer this, specifically: 
```c
#define VENDOR_ID       0x3297
#define PRODUCT_ID      0xC6CF
```

**Comment by mixedfeelings** [comment]
Would the olkb pcb's have the same VID as the EZ, or is that vendor ID unique to ZSA boards?

**Comment by drashna** [comment]
> Would the olkb pcb's have the same VID as the EZ, or is that vendor ID unique to ZSA boards?

The Vender ID is unique and belongs to ZSA.  It shouldn't be used for the OLKB boards (the EZ is a collaboration board between ZSA and OLKB/jack, but is sold and maintained by ZSA)

**Comment by rstacruz** [comment]
Hey all, I'd like to give this a try. Can I compile a firmware from this branch, and expect it to work with the VIA app?

**Comment by mixedfeelings** [comment]
> I assume this requires a .json file for Via with a keymap, any hints on how I can generate one?
 
Here you go: https://github.com/mixedfeelings/keyboards/tree/planck-rev6.1/src/olkb/planck/rev6

I haven't submitted a PR to VIA yet, as I am waiting for the blockers here to get resolved.

**Comment by jackhumbert** [comment]
Just to recap, the Planck EZ should remain untouched (it already has its VID/PID set) and these settings for `config.h` for the Planck Rev 1-5 will work for me:

```
#undefine PRODUCT_ID
#define PRODUCT_ID 0xAE01
```

the Planck Rev 6:

```
#undefine PRODUCT_ID
#define PRODUCT_ID 0xA4F9
```

and the Planck Light:

```
#undefine PRODUCT_ID
#define PRODUCT_ID 0xBEA2
```

I'd like to keep `0xFEED` as the Vendor ID for the above, if that's an option.

**Comment by mixedfeelings** [comment]
> I'd like to keep `0xFEED` as the Vendor ID for the above, if that's an option.

Thanks, @jackhumbert! Understood on all the above points, but unfortunately `0xFEED` can't be used for VIA. They require the Vendor ID to be unique and specific.

**Comment by jackhumbert** [comment]
Right! Some of that is coming back to me - let's use `0x03A8` for the VID. This was going to be the QMK-wide one at some point (with VIDs generated via command I mentioned earlier), but those plans haven't really been fleshed out yet.

**Comment by mixedfeelings** [comment]
You beat me too it, thank you so much!!

**Comment by drashna** [comment]
For reference, is there a Product ID to use for the Planck THK (#12597)?

**Comment by jackhumbert** [comment]
> For reference, is there a Product ID to use for the Planck THK (#12597)?

[Just added!](https://github.com/qmk/qmk_firmware/pull/12597/files#r617797937)

**Review by jackhumbert — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**